gio/src/cancellable_future.rs

```diff
@@ -32,14 +37,59 @@ pin_project! {
     /// c.cancel();
     ///
     /// ```
+    ///
+    /// As said the [`CancellableFuture`] can be used to handle Gio futures,
+    /// relying on an actual [`Cancellable`] instance or with a new one.
+    ///
+    /// ```no_run
+    /// # use futures::FutureExt;
+    /// # use gio::prelude::*;
+    /// # use gio::CancellableFuture;
+    /// # async {
+    /// CancellableFuture::new(
+    ///   async {
+    ///     let file = gio::File::for_path("/dev/null");
+    ///     let file_info = file
+    ///       .query_info_future(
+    ///         gio::FILE_ATTRIBUTE_STANDARD_NAME,
+    ///         gio::FileQueryInfoFlags::NONE,
+    ///         glib::Priority::default(),
+    ///       )
+    ///       .await?;
+    ///
+    ///     // Sub-cancellable chains are also working as expected.
+    ///     // The new cancellable will have its own scope, but will also be
+    ///     // cancelled when the parent cancellable is cancelled.
+    ///     let io_stream = CancellableFuture::new(
+    ///       file.open_readwrite_future(glib::Priority::default()),
+    ///       gio::Cancellable::new(),
+    ///     )
+    ///     .await??;
+    ///     // [...]
+    ///     Ok::<bool, glib::Error>(true)
+    ///   },
+    ///   gio::Cancellable::new(),
+    /// )
+    /// .await
+    /// # };
+    /// ```
     pub struct CancellableFuture<F> {
         #[pin]
         future: F,
 
-        #[pin]
-        waker_handler_cb: Option<CancelledHandlerId>,
-
         cancellable: Cancellable,
+        connection_id: Option<CancelledHandlerId>,
+
+        waker: std::sync::Arc<std::sync::Mutex<Option<std::task::Waker>>>,
```

Comment:
i'd avoid fullspec, `Arc` and `Mutex` can be imported to improve readability (no other `Arc` and `Mutex`, so there is no ambiguity)